doc/overviews/inference.md

```diff
@@ -0,0 +1,412 @@
+Then, they will never be set in single tenant model, or __only__ by the wasm-shim if running in multi tenant mode.
+
+__NOTE:__ At this time, the token based rate limiting feature is only available in multi tenant mode. This is because of how limiting is applied by limitador at the wasm-shim layer, not the inference service layer. The inference service is responsible for gathering and reporting on token usage stats, not the actual limiting.
+
+TODO: Add details on token rate limting implemenation after these issues are further along:
+
+* https://github.com/Kuadrant/kuadrant-operator/issues/1341
+* https://github.com/Kuadrant/kuadrant-operator/issues/1342
+* https://github.com/Kuadrant/kuadrant-operator/issues/1343
+
+### Wasm-shim configuration
+
+A Kuadrant wasm-shim configuration for one PromptGuardPolicy custom resources targeting a HTTPRoute looks like the following and is generated automatically by the Kuadrant control plane.
+Note the 2 different `services` in the `pluginConfig`, both using `ext_proc`.
+These both ultimately route to the same `cluster`, but on different ports.
+
+```yaml
+apiVersion: extensions.istio.io/v1alpha1
+kind: WasmPlugin
+metadata:
+  name: kuadrant-kuadrant-ingressgateway
+  namespace: gateway-system
+spec:
+  phase: STATS
+  pluginConfig:
+    services:
+      inference-promptguard-service:
+        type: ext_proc
+        endpoint: kuadrant-inference-promptguard-service
+        failureMode: deny
+      inference-tokenratelimit-service:
+        type: ext_proc
+        endpoint: kuadrant-inference-tokenratelimit-service
+        failureMode: allow
+    actionSets:
+      - name: some_name_0
+        routeRuleConditions:
+          hostnames:
+            - "*.models.website"
+            - "*.models.io"
+          predicates:
+            - request.url_path.startsWith("/openai/v1/completions")
+        actions:
+          - service: inference-promptguard-service
+            scope: gateway-system/app-promptguard
+            predicates:
+              - request.host.endsWith('.models.website')
+      - name: some_name_1
+        routeRuleConditions:
+          hostnames:
+            - "*.models.website"
+          predicates:
+            - request.url_path.startsWith("/openai/v1/completions")
+        actions:
+          - service: inference-tokenratelimit-service
+            scope: gateway-system/app-tokenratelimit
+            predicates:
+              - request.host.endsWith('.models.website')
+```
+
+Here is an example EnvoyFilter that configures the `cluster` for the wasm-shim to route to.
+Note the 2 different patches and ports going to the same `cluster` for the Inference service.
+The service is listening on different grpc ports for each feature to allow them to be executed at different points in the chain.
```

Comment:
@maleck13 when chatting in person you had some questions about the use of different ports and if there was a way to specify what feature the ext proc request should be executing for a specific request.

I think we *could* do something in the wasm-shim to hint at what feature should be executed e.g. via a header.
However, I'm not sure how you'd allow for that if the inference server was used standalone without the wasm-shim. Unless there's some way to set that context in the cluster patch below?

@jasonmadigan I suspect you've thought about this more before deciding on the ports approach.
Were there other options you came across?

I don't see any major issue with different ports, but I would be interested to know what the implementation on the inference service looks like for that?

Reply:
Would make it a bit tricker to run the component standalone, although at some point later if we do some perf testing, I guess we may want to be able to have options to scale the procs separately. Don't really see a reason to split now though, not onerous to do later if we absolutely need to?

> the ext proc request should be executing for a specific request

so with envoy standalone, we can add a `match` to a `virtual_hosts` `route`, then enable/disable a proc for that match via 

```yaml
typed_per_filter_config:
 envoy.filters.http.ext_proc:
   "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExtProcPerRoute
   disabled: true|false
```

think we can do similar with gateways/routes